{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.load import parse_paf, read_fai\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "paf = parse_paf('/data/home/sjwan/projects/Y-chromosome/workflow.output/03.assembly.workflow/RY02_Wei_Male_CNUGM0001/T2TY.minimap/ragtag.Y.paf')\n",
    "paf = paf.drop(paf[paf['target_name']=='*'].index)\n",
    "# paf = paf[paf['mapping_quality']>=60]\n",
    "fai = read_fai('/data/home/sjwan/projects/Y-chromosome/workflow.output/data/T2T_Y_subregion.fa.fai')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "contigs_name = paf['query_name'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# overlap_region = []\n",
    "# for contig in contigs_name:\n",
    "#     align = paf[paf['query_name']==contig]\n",
    "#     for _, row in align.iterrows():\n",
    "#         for _, region in fai.iterrows():\n",
    "#             if row['target_start'] <= region['start'] and row['target_end'] >= region['end']:\n",
    "#                 overlap_region.append([contig, region['contig']])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_usage(df, contig):\n",
    "    contig_df = df[df['query_name']==contig]\n",
    "    use_length = contig_df['query_end'] - contig_df['query_start']\n",
    "    return np.sum(use_length)/np.max(contig_df['query_length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_paf(paf):\n",
    "       keep_col = ['query_name', 'query_length', 'query_start', 'query_end', 'strand',\n",
    "              'target_name', 'target_length', 'target_start', 'target_end']\n",
    "       # paf = paf[paf['mapping_quality']>=60]\n",
    "       align = paf[keep_col]\n",
    "       align = align.sort_values(by=['query_start'],ascending=False)\n",
    "\n",
    "       align['start_diff'] =( align['query_start']/align['query_length']) - (align['target_start']/align['target_length'] )\n",
    "       align['end_diff'] = (align['query_end']/align['query_length']) - (align['target_end']/align['target_length']) \n",
    "       align['query_start_percent'] = align['query_start']/align['query_length']\n",
    "       align['query_end_percent'] = align['query_end']/align['query_length']\n",
    "       align['target_start_percent'] = align['target_start']/align['target_length']\n",
    "       align['target_end_percent'] = align['target_end']/align['target_length']\n",
    "       align['MB'] = (align['target_end'] - align['target_start'])/1e6\n",
    "       return align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib \n",
    "from collections import defaultdict\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "\n",
    "def plot_reads_cover(fai, plot_align, sample, out_fig):\n",
    "    plt.figure(figsize=(14, 8))\n",
    "    ref_colors = ListedColormap(plt.cm.tab20.colors[:len(fai)])\n",
    "    num_queries = plot_align['query_name'].nunique()\n",
    "    contig_total_length = plot_align.groupby('query_name')['MB'].sum().sort_values(ascending=False)\n",
    "    # 选择总体比对长度前五的 contig\n",
    "    top_contigs = contig_total_length.head(5).index\n",
    "    contig_colors = plt.cm.rainbow(np.linspace(0, 1, num_queries))\n",
    "    for i, row in fai.iterrows():\n",
    "        plt.plot([row['start'], row['end']], [-1000000, -1000000], marker='|', color=ref_colors(i))\n",
    "    # 动态分配 y 轴位置\n",
    "    y_positions = defaultdict(list)\n",
    "    max_y = 15  # 初始最大 y 位置\n",
    "    query_names = plot_align['query_name'].unique()\n",
    "    query_color_map = {query_name: contig_colors[i] for i, query_name in enumerate(query_names)}\n",
    "    # 分配给每个 query_name 的颜色\n",
    "    query_names = plot_align['query_name'].unique()\n",
    "    query_color_map = {query_name: contig_colors[i] for i, query_name in enumerate(query_names)}\n",
    "    # 颜色映射\n",
    "    top_colors = plt.cm.tab10(np.linspace(0, 1, 5))  # 前五个 contigs 使用 tab10 的前五种颜色\n",
    "    other_colors = plt.cm.rainbow(np.linspace(0, 1, max(len(plot_align['query_name'].unique()) - 5, 0)))  # 其他 contigs 使用 rainbow 颜色\n",
    "    query_color_map = {}\n",
    "    for i, query_name in enumerate(plot_align['query_name'].unique()):\n",
    "        if query_name in top_contigs:\n",
    "            query_color_map[query_name] = top_colors[list(top_contigs).index(query_name)]\n",
    "        else:\n",
    "            query_color_map[query_name] = other_colors[len(query_color_map) - len(top_contigs)]\n",
    "\n",
    "    # 绘制比对结果\n",
    "    for i, row in plot_align.iterrows():\n",
    "        contig_color = query_color_map[row['query_name']]\n",
    "        query_start = row['query_start']\n",
    "        query_end = row['query_end']\n",
    "        if row['strand'] == '-':\n",
    "            query_start, query_end = query_end, query_start  # 反转方向\n",
    "        plt.plot([row['target_start'], row['target_end']], [query_start, query_end], marker='o', color=contig_color,\n",
    "                 label=row['query_name'] if row['query_name'] in top_contigs else \"\")\n",
    "\n",
    "    # 添加图例\n",
    "    handles, labels = plt.gca().get_legend_handles_labels()\n",
    "    by_label = dict(zip(labels, handles))\n",
    "    # 只保留前五个 contig 的图例\n",
    "    top_handles = [by_label[label] for label in top_contigs]\n",
    "    top_labels = top_contigs\n",
    "    plt.legend(top_handles, top_labels, title=\"Top 5 Contigs by Alignment Length\")\n",
    "\n",
    "    # 调整 y 轴范围以适应所有比对结果\n",
    "    plt.ylim(-2000000, plot_align[['query_start', 'query_end']].max().max() + 2000000)\n",
    "    plt.xlim(-2000000, plot_align['target_end'].max() + 2000000)\n",
    "\n",
    "    # 调整 y 轴范围以适应所有比对结果\n",
    "    # plt.ylim(-1, max_y + 1)\n",
    "    # 添加标题和标签\n",
    "    plt.title(f'Dotplot of {sample} Minimap Alignments')\n",
    "    plt.xlabel('Position')\n",
    "    plt.ylabel('Alignments')\n",
    "    plt.grid(True)\n",
    "    plt.savefig(out_fig)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "data_dir = '/data/home/sjwan/projects/Y-chromosome/workflow.output/data/verkko1.4'\n",
    "sample_list = os.listdir(data_dir)\n",
    "sample_list = sorted(sample_list)\n",
    "# for sample in sample_list:\n",
    "#     out_fig = f'/data/home/sjwan/projects/Y-chromosome/workflow.output/03.assembly.workflow/img/minimap.Y.{sample}.png'\n",
    "#     paf = parse_paf(f'/data/home/sjwan/projects/Y-chromosome/workflow.output/03.assembly.workflow/{sample}/T2TY.minimap/ragtag.Y.paf')\n",
    "#     plot_align = filter_paf(paf)\n",
    "#     plot_align = plot_align[plot_align['MB']>=0.2].sort_values(by=['MB'], ascending=False)\n",
    "#     plot_reads_cover(fai, plot_align, sample,out_fig)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep_contigs = {}\n",
    "for sample in sample_list:\n",
    "    paf = parse_paf(f'/data/home/sjwan/projects/Y-chromosome/workflow.output/03.assembly.workflow/{sample}/T2TY.minimap/ragtag.Y.paf')\n",
    "    plot_align = filter_paf(paf)\n",
    "    contig_total_length = plot_align.groupby('query_name')['MB'].sum().sort_values(ascending=False)\n",
    "    top_contigs = contig_total_length.head(10).index\n",
    "    keep_contigs[sample] = list(top_contigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.fasta import extract_contigs\n",
    "for sample in sample_list:\n",
    "    contigs = f'/data/home/sjwan/projects/Y-chromosome/workflow.output/data/verkko1.4/{sample}/contigs.ragtag.Y.fasta'\n",
    "    out_contigs = f'/data/home/sjwan/projects/Y-chromosome/workflow.output/data/verkko1.4/{sample}/contigs.max10.Y.fasta'\n",
    "    extract_contigs(contigs, out_contigs,keep_contigs[sample])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_name = fai['contig'].unique().tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2'], ['unassigned-0002298', 'chrY_XDR2'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR3'], ['unassigned-0002298', 'chrY_AMPL3'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_AMPL5'], ['unassigned-0002298', 'chrY_XDR6'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_HET2_DYZ19'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other2_DYZ18']]\n",
      "[['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other2_DYZ18'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_HET2_DYZ19'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_XDR6'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centroDYZ3'], ['unassigned-0002298', 'chrY_HET1_centro2'], ['unassigned-0002298', 'chrY_XDR3'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2'], ['unassigned-0002298', 'chrY_XDR2'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1']]\n",
      "[['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_XDR3'], ['unassigned-0002298', 'chrY_AMPL3'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_AMPL5'], ['unassigned-0002298', 'chrY_XDR6'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centroDYZ3'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2'], ['unassigned-0002298', 'chrY_XDR2'], ['unassigned-0002298', 'chrY_other2_DYZ18']]\n",
      "[['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2'], ['unassigned-0002298', 'chrY_XDR2'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centroDYZ3'], ['unassigned-0002298', 'chrY_other2_DYZ18'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR3'], ['unassigned-0002298', 'chrY_AMPL3'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_AMPL5'], ['unassigned-0002298', 'chrY_XDR6']]\n",
      "[['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2'], ['unassigned-0002298', 'chrY_XDR2'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centroDYZ3'], ['unassigned-0002298', 'chrY_XDR3'], ['unassigned-0002298', 'chrY_AMPL3'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_AMPL5'], ['unassigned-0002298', 'chrY_XDR6'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_other2_DYZ18']]\n",
      "[['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other2_DYZ18'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_HET2_DYZ19'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_AMPL5'], ['unassigned-0002298', 'chrY_XDR6'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2'], ['unassigned-0002298', 'chrY_XDR2'], ['unassigned-0002298', 'chrY_AMPL2'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro2'], ['unassigned-0002298', 'chrY_XDR3']]\n",
      "[['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_other2_DYZ18'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR3'], ['unassigned-0002298', 'chrY_AMPL3'], ['unassigned-0002298', 'chrY_XDR4'], ['unassigned-0002298', 'chrY_AMPL4'], ['unassigned-0002298', 'chrY_XDR5'], ['unassigned-0002298', 'chrY_AMPL5'], ['unassigned-0002298', 'chrY_XDR6'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR7'], ['unassigned-0002298', 'chrY_XDR8'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_AMPL2'], ['unassigned-0002298', 'chrY_other1'], ['unassigned-0002298', 'chrY_HET1_centro1'], ['unassigned-0002298', 'chrY_XDR1'], ['unassigned-0002298', 'chrY_XTR1'], ['unassigned-0002298', 'chrY_AMPL1'], ['unassigned-0002298', 'chrY_XTR2']]\n"
     ]
    }
   ],
   "source": [
    "all_count = {}\n",
    "for sample in sample_list:\n",
    "    overlap_region = []\n",
    "    paf = parse_paf(f'/data/home/sjwan/projects/Y-chromosome/workflow.output/03.assembly.workflow/{sample}/T2TY.minimap/ragtag.Y.paf')\n",
    "    plot_align = filter_paf(paf)\n",
    "    for idx, row in plot_align.iterrows():\n",
    "        for _, region in fai.iterrows():\n",
    "            if row['target_start'] <= region['start'] and row['target_end'] >= region['end']:\n",
    "                overlap_region.append([contig, region['contig']])\n",
    "    all_count[sample] = overlap_region\n",
    "    print(overlap_region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_name = [name.split('_')[1] for name in regions_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "continue_array = np.zeros([len(sample_list), len(regions_name)])\n",
    "continue_df = pd.DataFrame(continue_array, columns=regions_name)\n",
    "for idx, sample in enumerate(sample_list):\n",
    "    regions = all_count[sample]\n",
    "    for region in regions:\n",
    "        continue_df.loc[idx,region[1].split('_')[1]] = 1\n",
    "continue_df.index = sample_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "continue_df.to_csv('assemblies.regions.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_gap(cover_array):\n",
    "    gaps = []\n",
    "    start = 0\n",
    "    end = 0\n",
    "    while start <= cover_array.shape[0]:\n",
    "        if cover_array[end] == 1:\n",
    "            end+=1\n",
    "            \n",
    "\n",
    "def count_contigs_coverage(paf_df):\n",
    "    unique_name = paf_df['query_name'].unique()\n",
    "    line = []\n",
    "    for name in unique_name:\n",
    "        part_df = paf_df[paf_df['query_name']==name]\n",
    "        cover = np.zeros(part_df['query_length'].max()+1,dtype=np.int8)\n",
    "        min_start = part_df['target_start'].min()\n",
    "        max_end = part_df['target_end'].max()\n",
    "        for idx, row in part_df.iterrows():\n",
    "            cover[int(row['query_start']):int(row['query_end'])] = 1\n",
    "        ref_coverage = (max_end - min_start)/part_df['query_length'].max()\n",
    "        line.append([name, part_df['query_length'].max(),np.mean(cover), min_start, max_end, ref_coverage])\n",
    "    return pd.DataFrame(line, columns=['contig','length','cover','ref_start','ref_end','ref_coverage'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "cover_df =  count_contigs_coverage(paf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>length</th>\n",
       "      <th>cover</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>ref_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>haplotype2-0000666</td>\n",
       "      <td>54977839</td>\n",
       "      <td>0.969785</td>\n",
       "      <td>0</td>\n",
       "      <td>62460029</td>\n",
       "      <td>1.136095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>haplotype1-0000031</td>\n",
       "      <td>974225</td>\n",
       "      <td>0.993788</td>\n",
       "      <td>0</td>\n",
       "      <td>1009835</td>\n",
       "      <td>1.036552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>haplotype2-0000686</td>\n",
       "      <td>280319</td>\n",
       "      <td>0.999633</td>\n",
       "      <td>1085253</td>\n",
       "      <td>1371467</td>\n",
       "      <td>1.021030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>haplotype2-0000687</td>\n",
       "      <td>281585</td>\n",
       "      <td>0.999634</td>\n",
       "      <td>1086871</td>\n",
       "      <td>1371045</td>\n",
       "      <td>1.009194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>haplotype2-0000732</td>\n",
       "      <td>130585</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>1410286</td>\n",
       "      <td>1540242</td>\n",
       "      <td>0.995183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>haplotype2-0000722</td>\n",
       "      <td>342875</td>\n",
       "      <td>0.999519</td>\n",
       "      <td>1543922</td>\n",
       "      <td>1889038</td>\n",
       "      <td>1.006536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>haplotype2-0000723</td>\n",
       "      <td>342486</td>\n",
       "      <td>0.999936</td>\n",
       "      <td>1545848</td>\n",
       "      <td>1888377</td>\n",
       "      <td>1.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>haplotype2-0000724</td>\n",
       "      <td>108339</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>1880835</td>\n",
       "      <td>1989462</td>\n",
       "      <td>1.002658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>haplotype2-0000734</td>\n",
       "      <td>294652</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>10570352</td>\n",
       "      <td>10872152</td>\n",
       "      <td>1.024259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>haplotype2-0000715</td>\n",
       "      <td>35110</td>\n",
       "      <td>0.999972</td>\n",
       "      <td>10571983</td>\n",
       "      <td>10717006</td>\n",
       "      <td>4.130533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>haplotype2-0000716</td>\n",
       "      <td>44005</td>\n",
       "      <td>0.999977</td>\n",
       "      <td>10572240</td>\n",
       "      <td>10622030</td>\n",
       "      <td>1.131462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>haplotype2-0000717</td>\n",
       "      <td>121900</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>10573154</td>\n",
       "      <td>10700835</td>\n",
       "      <td>1.047424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>unassigned-0001366</td>\n",
       "      <td>170715</td>\n",
       "      <td>0.840829</td>\n",
       "      <td>11250874</td>\n",
       "      <td>62072780</td>\n",
       "      <td>297.700296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>unassigned-0001362</td>\n",
       "      <td>185122</td>\n",
       "      <td>0.850856</td>\n",
       "      <td>11250874</td>\n",
       "      <td>62072780</td>\n",
       "      <td>274.531963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>unassigned-0001381</td>\n",
       "      <td>53085</td>\n",
       "      <td>0.999774</td>\n",
       "      <td>11344274</td>\n",
       "      <td>11394843</td>\n",
       "      <td>0.952604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>unassigned-0001402</td>\n",
       "      <td>40772</td>\n",
       "      <td>0.999975</td>\n",
       "      <td>11344460</td>\n",
       "      <td>11382745</td>\n",
       "      <td>0.939002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>unassigned-0001382</td>\n",
       "      <td>29552</td>\n",
       "      <td>0.999966</td>\n",
       "      <td>11365883</td>\n",
       "      <td>11395422</td>\n",
       "      <td>0.999560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511</th>\n",
       "      <td>unassigned-0001573</td>\n",
       "      <td>45289</td>\n",
       "      <td>0.854560</td>\n",
       "      <td>11525534</td>\n",
       "      <td>11548072</td>\n",
       "      <td>0.497648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>unassigned-0001572</td>\n",
       "      <td>46797</td>\n",
       "      <td>0.871533</td>\n",
       "      <td>11525534</td>\n",
       "      <td>11548072</td>\n",
       "      <td>0.481612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>unassigned-0001264</td>\n",
       "      <td>67750</td>\n",
       "      <td>0.917433</td>\n",
       "      <td>11525780</td>\n",
       "      <td>11552515</td>\n",
       "      <td>0.394613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>unassigned-0001265</td>\n",
       "      <td>66804</td>\n",
       "      <td>0.925859</td>\n",
       "      <td>11525780</td>\n",
       "      <td>11555784</td>\n",
       "      <td>0.449135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>haplotype2-0000754</td>\n",
       "      <td>162148</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>57283988</td>\n",
       "      <td>59391574</td>\n",
       "      <td>12.997915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>haplotype2-0000755</td>\n",
       "      <td>42066</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>57314044</td>\n",
       "      <td>58420124</td>\n",
       "      <td>26.293919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>unassigned-0001100</td>\n",
       "      <td>30398</td>\n",
       "      <td>0.999934</td>\n",
       "      <td>62091318</td>\n",
       "      <td>62120805</td>\n",
       "      <td>0.970031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>unassigned-0001101</td>\n",
       "      <td>32939</td>\n",
       "      <td>0.919460</td>\n",
       "      <td>62093323</td>\n",
       "      <td>62122781</td>\n",
       "      <td>0.894320</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 contig    length     cover  ref_start   ref_end  ref_coverage\n",
       "48   haplotype2-0000666  54977839  0.969785          0  62460029      1.136095\n",
       "10   haplotype1-0000031    974225  0.993788          0   1009835      1.036552\n",
       "59   haplotype2-0000686    280319  0.999633    1085253   1371467      1.021030\n",
       "60   haplotype2-0000687    281585  0.999634    1086871   1371045      1.009194\n",
       "74   haplotype2-0000732    130585  0.999992    1410286   1540242      0.995183\n",
       "69   haplotype2-0000722    342875  0.999519    1543922   1889038      1.006536\n",
       "70   haplotype2-0000723    342486  0.999936    1545848   1888377      1.000126\n",
       "71   haplotype2-0000724    108339  0.999991    1880835   1989462      1.002658\n",
       "76   haplotype2-0000734    294652  0.999997   10570352  10872152      1.024259\n",
       "65   haplotype2-0000715     35110  0.999972   10571983  10717006      4.130533\n",
       "66   haplotype2-0000716     44005  0.999977   10572240  10622030      1.131462\n",
       "67   haplotype2-0000717    121900  0.999992   10573154  10700835      1.047424\n",
       "403  unassigned-0001366    170715  0.840829   11250874  62072780    297.700296\n",
       "400  unassigned-0001362    185122  0.850856   11250874  62072780    274.531963\n",
       "409  unassigned-0001381     53085  0.999774   11344274  11394843      0.952604\n",
       "426  unassigned-0001402     40772  0.999975   11344460  11382745      0.939002\n",
       "410  unassigned-0001382     29552  0.999966   11365883  11395422      0.999560\n",
       "511  unassigned-0001573     45289  0.854560   11525534  11548072      0.497648\n",
       "510  unassigned-0001572     46797  0.871533   11525534  11548072      0.481612\n",
       "332  unassigned-0001264     67750  0.917433   11525780  11552515      0.394613\n",
       "333  unassigned-0001265     66804  0.925859   11525780  11555784      0.449135\n",
       "83   haplotype2-0000754    162148  0.999994   57283988  59391574     12.997915\n",
       "84   haplotype2-0000755     42066  0.999976   57314044  58420124     26.293919\n",
       "260  unassigned-0001100     30398  0.999934   62091318  62120805      0.970031\n",
       "261  unassigned-0001101     32939  0.919460   62093323  62122781      0.894320"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cover_df[cover_df['cover']>=0.8].sort_values(by=['length'],ascending=False).head(25).sort_values(by=['ref_start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.cluster import DBSCAN\n",
    "from matplotlib.colors import Normalize\n",
    "\n",
    "# 计算偏移量 b，考虑正链和反链的情况\n",
    "def calculate_b(row):\n",
    "    try:\n",
    "        if row['strand'] == '+':\n",
    "            return (row['target_start']/row['target_length'] - row['query_start']/row['query_length'])\n",
    "        else:\n",
    "            return (row['target_end']/row['target_length'] - row['query_start']/row['query_length'])\n",
    "    except ZeroDivisionError:\n",
    "        print(row)\n",
    "\n",
    "def calculate_d(row):\n",
    "    try:\n",
    "        if row['strand'] == '+':\n",
    "            return (row['target_end']/row['target_length'] - row['query_end']/row['query_length'])\n",
    "        else:\n",
    "            return (row['target_start']/row['target_length'] - row['query_end']/row['query_length'])\n",
    "    except ZeroDivisionError:\n",
    "        print(row)\n",
    "    \n",
    "    \n",
    "def cluster_most_coverage(paf_df, contig_names):\n",
    "    line = []\n",
    "    for name in contig_names:\n",
    "        part_df = paf_df[paf_df['query_name'] == name].copy()\n",
    "        part_df['b'] = part_df.apply(calculate_b, axis=1)\n",
    "        cover = np.zeros(part_df['query_length'].max() + 1, dtype=np.int8)\n",
    "        min_start = part_df['target_start'].min()\n",
    "        max_end = part_df['target_end'].max()\n",
    "        for idx, row in part_df.iterrows():\n",
    "            cover[int(row['query_start']):int(row['query_end'])] = 1\n",
    "        if np.mean(cover) < 0.9 and part_df['query_length'].max() <= 1000000:\n",
    "            continue\n",
    "        \n",
    "        # 使用DBSCAN进行聚类\n",
    "        eps = 0.005  # 需要根据数据特点调整\n",
    "        min_samples = 2  # 需要根据数据特点调整\n",
    "        dbscan = DBSCAN(eps=eps, min_samples=min_samples)\n",
    "        part_df.loc[part_df.index, 'cluster'] = dbscan.fit_predict(part_df[['b']])\n",
    "        \n",
    "        print(name, part_df['cluster'].unique())\n",
    "        if (np.sum(part_df['cluster'] == -1) / part_df.shape[0]) >= 0.5:\n",
    "            continue\n",
    "        # part_df['b'].hist()\n",
    "        fig, ax = plt.subplots(figsize=(10, 6))\n",
    "        \n",
    "        # 定义多个色系\n",
    "        color_maps = [plt.cm.viridis, plt.cm.plasma, plt.cm.inferno, plt.cm.magma, plt.cm.cividis]\n",
    "        cluster_color_map = {}\n",
    "        cluster_lines = {}  # 用于保存每个簇的Line2D对象，以便在图例中显示\n",
    "\n",
    "        cluster_ids = sorted(part_df['cluster'].unique())\n",
    "        for i, cluster in enumerate(cluster_ids):\n",
    "            cluster_color_map[cluster] = color_maps[i % len(color_maps)]\n",
    "        \n",
    "        for idx, row in part_df.iterrows():\n",
    "            cluster = row['cluster']\n",
    "            cmap = cluster_color_map[cluster]\n",
    "            color = cmap(0.5)  # 取颜色图的中间颜色\n",
    "            line_style = 'solid' if row['strand'] == '+' else 'dashed'\n",
    "            line, = plt.plot(\n",
    "                [row['query_start'], row['query_end']], \n",
    "                [row['target_start'], row['target_end']] if row['strand'] == '+' else [row['target_end'], row['target_start']], \n",
    "                color=color, linestyle=line_style,\n",
    "                label=f'Cluster {cluster}' if cluster not in cluster_lines else \"\"\n",
    "            )\n",
    "            if cluster not in cluster_lines:\n",
    "                cluster_lines[cluster] = line\n",
    "        \n",
    "        # 添加图例\n",
    "        handles, labels = [], []\n",
    "        for cluster, line in cluster_lines.items():\n",
    "            handles.append(line)\n",
    "            labels.append(f'Cluster {cluster}')\n",
    "        ax.legend(handles=handles, labels=labels, title='Clusters', loc='upper right')\n",
    "        \n",
    "        ax.set_xlabel('Query Start')\n",
    "        ax.set_ylabel('Target Start')\n",
    "        ax.set_title('DBSCAN Clustering of Alignments Based on Offset b')\n",
    "        plt.savefig(f'/data/home/sjwan/projects/Y-chromosome/tmp/{name}.png')\n",
    "        plt.close()\n",
    "        # return None\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "haplotype1-0000004 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. -1. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52.\n",
      " 53.]\n",
      "haplotype1-0000007 [-1.  0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48.]\n",
      "haplotype1-0000009 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. -1. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52.]\n",
      "haplotype1-0000011 [ 0.  1.  2. -1.  3.  4.  5.  6.  7.  8.  9.]\n",
      "haplotype1-0000017 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.\n",
      " 18. 19. 20. 21. 22. 23. 24. 25. 26. -1. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50.]\n",
      "haplotype1-0000019 [ 0. -1.  1.  2.  3.]\n",
      "haplotype1-0000020 [ 0.  1.  2.  3.  4.  5.  6.  7. -1.  8.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38.]\n",
      "haplotype1-0000026 [ 0.  1.  2.  3.  4.  5.  6.  7.  8. -1.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46.]\n",
      "haplotype1-0000027 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. -1. 13. 14.]\n",
      "haplotype1-0000030 [ 0. -1.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14.]\n",
      "haplotype1-0000031 [-1.  0.]\n",
      "haplotype1-0000032 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.\n",
      " 18. 19. 20. -1. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51.]\n",
      "haplotype1-0000033 [ 0.  1.  2.  3. -1.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19.]\n",
      "haplotype1-0000034 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. -1. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50.]\n",
      "haplotype1-0000035 [ 0.  1.  2.  3. -1.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42.]\n",
      "haplotype2-0000651 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. -1.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42.]\n",
      "haplotype2-0000653 [ 0.  1.  2.  3.  4.  5. -1.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52.\n",
      " 53. 54. 55. 56.]\n",
      "haplotype2-0000666 [ 0.  1.  2.  3.  4.  5.  6. -1.  7.  8.  9. 10. 11. 12.]\n",
      "haplotype2-0000668 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. -1. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50.]\n",
      "haplotype2-0000669 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. -1. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52.\n",
      " 53. 54. 55. 56. 57. 58. 59. 60. 61. 62.]\n",
      "haplotype2-0000670 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.\n",
      " -1. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52.\n",
      " 53. 54. 55. 56.]\n",
      "haplotype2-0000671 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. -1. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52.\n",
      " 53.]\n",
      "haplotype2-0000673 [-1.  0.  1.]\n",
      "haplotype2-0000678 [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.\n",
      " 18. 19. 20. 21. -1. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48.]\n",
      "haplotype2-0000681 [ 0.  1.  2.  3.  4.  5.  6.  7.  8. -1.  9. 10. 11. 12. 13. 14. 15. 16.\n",
      " 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34.\n",
      " 35. 36. 37. 38. 39. 40. 41.]\n",
      "haplotype2-0000686 [0.]\n",
      "haplotype2-0000687 [0.]\n",
      "haplotype2-0000715 [0.]\n",
      "haplotype2-0000716 [0.]\n",
      "haplotype2-0000717 [0.]\n",
      "haplotype2-0000718 [0.]\n",
      "haplotype2-0000722 [0.]\n",
      "haplotype2-0000723 [-1.]\n",
      "haplotype2-0000724 [-1.]\n",
      "haplotype2-0000732 [-1.]\n",
      "haplotype2-0000734 [0.]\n",
      "haplotype2-0000754 [-1.]\n",
      "haplotype2-0000755 [-1.]\n",
      "unassigned-0000769 [-1.]\n",
      "unassigned-0000912 [ 0.  1. -1.  2.  3.  4.  5.  6.  7.  8.]\n",
      "unassigned-0000947 [-1.]\n",
      "unassigned-0001015 [-1.]\n",
      "unassigned-0001100 [-1.]\n",
      "unassigned-0001101 [-1.]\n",
      "unassigned-0001104 [-1.]\n",
      "unassigned-0001264 [0.]\n",
      "unassigned-0001265 [0.]\n",
      "unassigned-0001321 [-1.]\n",
      "unassigned-0001322 [-1.]\n",
      "unassigned-0001381 [-1.]\n",
      "unassigned-0001382 [-1.]\n",
      "unassigned-0001385 [-1.]\n",
      "unassigned-0001402 [-1.]\n",
      "unassigned-0002014 [-1.]\n",
      "unassigned-0002565 [0.]\n",
      "unassigned-0002630 [-1.]\n"
     ]
    }
   ],
   "source": [
    "cluster_most_coverage(paf,cover_df['contig'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# paf[paf['target_name']=='*']\n",
    "df_clean = paf.drop(paf[paf['target_name']=='*'].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_all_diff(paf_df):\n",
    "    part_df = paf_df.copy()\n",
    "    part_df['b'] = part_df.apply(calculate_b, axis=1)\n",
    "    part_df['d'] = part_df.apply(calculate_d, axis=1)\n",
    "    # 使用DBSCAN进行聚类\n",
    "    eps = 0.05  # 需要根据数据特点调整\n",
    "    min_samples = 3  # 需要根据数据特点调整\n",
    "    dbscan = DBSCAN(eps=eps, min_samples=min_samples)\n",
    "    part_df.loc[part_df.index, 'cluster'] = dbscan.fit_predict(part_df[['b','d']])\n",
    "    # part_df['b'].hist()\n",
    "    fig, ax = plt.subplots(figsize=(10, 6))\n",
    "\n",
    "    # 定义多个色系\n",
    "    color_maps = [plt.cm.viridis, plt.cm.plasma, plt.cm.inferno, plt.cm.magma, plt.cm.cividis]\n",
    "    cluster_color_map = {}\n",
    "    cluster_lines = {}  # 用于保存每个簇的Line2D对象，以便在图例中显示\n",
    "\n",
    "    cluster_ids = sorted(part_df['cluster'].unique())\n",
    "    for i, cluster in enumerate(cluster_ids):\n",
    "        cluster_color_map[cluster] = color_maps[i % len(color_maps)]\n",
    "\n",
    "    for idx, row in part_df.iterrows():\n",
    "        cluster = row['cluster']\n",
    "        cmap = cluster_color_map[cluster]\n",
    "        color = cmap(0.5)  # 取颜色图的中间颜色\n",
    "        line_style = 'solid' if row['strand'] == '+' else 'dashed'\n",
    "        line, = plt.plot(\n",
    "            [row['query_start'], row['query_end']],\n",
    "            [row['target_start'], row['target_end']] if row['strand'] == '+' else [row['target_end'], row['target_start']],\n",
    "            color=color, linestyle=line_style,\n",
    "            label=f'Cluster {cluster}' if cluster not in cluster_lines else \"\"\n",
    "        )\n",
    "        if cluster not in cluster_lines:\n",
    "            cluster_lines[cluster] = line\n",
    "\n",
    "    # 添加图例\n",
    "    handles, labels = [], []\n",
    "    for cluster, line in cluster_lines.items():\n",
    "        handles.append(line)\n",
    "        labels.append(f'Cluster {cluster}')\n",
    "    ax.legend(handles=handles, labels=labels, title='Clusters', loc='upper right')\n",
    "\n",
    "    ax.set_xlabel('Query Start')\n",
    "    ax.set_ylabel('Target Start')\n",
    "    ax.set_title('DBSCAN Clustering of Alignments Based on Offset b')\n",
    "    plt.savefig(f'/data/home/sjwan/projects/Y-chromosome/tmp/all.png')\n",
    "    plt.close()\n",
    "    return part_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = cluster_all_diff(paf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<Axes: title={'center': 'b'}>, <Axes: title={'center': 'd'}>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAikAAAGzCAYAAADqhoemAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8WgzjOAAAACXBIWXMAAA9hAAAPYQGoP6dpAAArQklEQVR4nO3df3RU5Z3H8U8SkglBhjQICVkDidbKb0JBQtTVVmOipl2wbC02q5FloYvBFbOLhj3yI6AGqUs50Ci6qyBHqdazghYQSKOVVULAiBWBUmhB3NIJtRTCDwlD5u4f7kwdQkgmmZn7zMz7dQ4H5s5z73yfzMyXzzxzZxJnWZYlAAAAw8TbXQAAAMDFEFIAAICRCCkAAMBIhBQAAGAkQgoAADASIQUAABiJkAIAAIxESAEAAEYipAAAACMRUhAS8+bNU1xcnD7//HO7SwEQQ7y9B9GBkAIAAIxESAEAAEYipAAAACMRUhBSn3/+ue666y45nU717t1bDz74oM6ePWt3WQCiwHvvvadrr71WycnJuuqqq/Tss8/aXRKCrJvdBSC63XXXXcrOzlZVVZW2bdumpUuX6i9/+YtWrVpld2kAItiuXbtUWFioPn36aN68eTp//rzmzp2r9PR0u0tDEBFSEFI5OTl64403JEllZWVyOp16+umn9W//9m8aPny4zdUBiFRz5syRZVn6n//5H/Xv31+SNGHCBA0bNszmyhBMvN2DkCorK/O7/MADD0iSNmzYYEc5AKJAS0uLNm3apPHjx/sCiiQNGjRIRUVFNlaGYCOkIKSuvvpqv8tXXXWV4uPjdejQIXsKAhDx/vSnP+mLL75o1V8k6ZprrrGhIoQKIQVhxZcsAQA6ipCCkNq/f7/f5QMHDsjj8Sg7O9ueggBEvD59+qh79+6t+osk7du3z4aKECqEFIRUdXW13+Vly5ZJkm6//XY7ygEQBRISElRUVKS1a9fq8OHDvu179+7Vpk2bbKwMwcanexBSBw8e1N/93d/ptttuU11dnV566SX98Ic/1IgRI+wuDUAEq6ys1MaNG/W3f/u3uv/++3X+/HktW7ZMQ4YM0ccff2x3eQgSVlIQUq+++qocDocqKiq0fv16TZ8+Xc8//7zdZQGIcMOHD9emTZvUp08fzZkzRy+88IIqKyt155132l0agijOsizL7iIAAAAuxEoKAAAwEiEFAAAYiZACAACMREgBAABGIqQAAAAjEVIAAICRIvLL3Dwej44cOaKePXvyu2CAILMsSydPnlRmZqbi42PvdQz9BQidQPtLRIaUI0eOKCsry+4ygKj22Wef6YorrrC7jLCjvwCh19H+EpEhpWfPnpK+nKTT6WxznNvt1ubNm1VYWKjExMRwlRdUzMEMsTSHpqYmZWVl+Z5nsYb+ElmYgxlC1V8iMqR4l2CdTme7TSQlJUVOpzOi73jmYL9YnEOsvtVBf4kszMEMoeovsfeGMwAAiAiEFAAAYCRCCgAAMBIhBQAAGImQAgAAjERIAQAARiKkAAAAIxFSAACAkQgpAADASIQUAABgJEIKAAAwEiEFAAAYiZACAACMREgBAABGIqTEgOyK9cquWG93GQCiEP0FoURIAQAARiKkAAAAIxFSAACAkQgpAADASIQUAABgJEIKAAAwEiEFAAAYiZACAACMREgBAABGIqQAAAAjEVIAAICRCCkAAMBIhBQAAGAkQgoAADASIQUAABiJkAIAAIxESAEAAEYipAAAACMRUgAAgJEIKQAAwEiEFIRVdsV6ZVest7sMAEAEIKQAAAAjEVIAABFv6LxNdpeAECCkIGxoIgCAQBBSAACAkQgpAICowIn50YeQAgAAjERIAWCElpYWzZ49Wzk5OerevbuuuuoqLViwQJZl+cZYlqU5c+aoX79+6t69uwoKCrR//36/4xw7dkwlJSVyOp1KTU3V5MmTderUqXBPB0AQBBRSaCIAQuXJJ5/UM888o5/+9Kfau3evnnzySS1atEjLli3zjVm0aJGWLl2q5cuXq76+Xj169FBRUZHOnj3rG1NSUqLdu3erpqZG69at05YtWzR16lQ7pgSgiwIKKTQRAKGydetWjRs3TsXFxcrOztbf//3fq7CwUNu3b5f05QugJUuW6NFHH9W4ceM0fPhwrVq1SkeOHNHatWslSXv37tXGjRv1X//1X8rLy9MNN9ygZcuW6ZVXXtGRI0dsnB2AzugWyOCvNhFJys7O1s9+9rM2m4gkrVq1Sunp6Vq7dq0mTpzoayI7duzQ6NGjJUnLli3THXfcoaeeekqZmZnBnB+ACHHdddfpueee029/+1t94xvf0K9//Wu99957Wrx4sSTp4MGDcrlcKigo8O3Tq1cv5eXlqa6uThMnTlRdXZ1SU1N9vUWSCgoKFB8fr/r6et15552tbre5uVnNzc2+y01NTZIkt9stt9vdZr3e6y41xnTBmIMjweryMbrCe7uOeKvVtkgRS4+lQOcYUEihiYQfTcQMsfRYsmuOFRUVampq0sCBA5WQkKCWlhY9/vjjKikpkSS5XC5JUnp6ut9+6enpvutcLpf69u3rd323bt2UlpbmG3OhqqoqVVZWttq+efNmpaSktFt3TU1N+5MzXFfmsGjMl39v2LAhSNV0zoLRHt+/7a6ls2LhsXTmzJmAjhdQSKGJ2IcmYoZYeCwF2kSC5ec//7lefvllrV69WkOGDNFHH32kGTNmKDMzU6WlpSG73VmzZqm8vNx3uampSVlZWSosLJTT6WxzP7fbrZqaGt16661KTEwMWX2hFIw5eL+k8ZN5RZ26vqu8c5j9QbyaPXEhva1QiaXHkneRoaMCCik0kfALRxMJNZqIGULVRIJl5syZqqio0MSJEyVJw4YN06effqqqqiqVlpYqIyNDktTY2Kh+/fr59mtsbFRubq4kKSMjQ0ePHvU77vnz53Xs2DHf/hdyOBxyOBytticmJnbovu7oOJNdbA7e7xs5tLD4kvs2t8T5jtGZ64Ol2RMXttsKlWh9LF14fSACCik0Eft0ZQ7heOJ2pKHRRMwQ7CYSLGfOnFF8vP+5/AkJCfJ4vlyBy8nJUUZGhmpra339pKmpSfX19Zo2bZokKT8/X8ePH1dDQ4NGjRolSXr77bfl8XiUl5cXvskACIqAPt0TSBPx8jaR/Px8Sf5NxIsmYh++oRGm+O53v6vHH39c69ev16FDh7RmzRotXrzYd55aXFycZsyYoccee0xvvvmmdu3apXvvvVeZmZkaP368JGnQoEG67bbbNGXKFG3fvl3vv/++pk+frokTJ3JSPhCBAlpJ8TaR/v37a8iQIdq5c6cWL16sf/zHf5Tk30Suvvpq5eTkaPbs2W02keXLl8vtdtNEAGjZsmWaPXu27r//fh09elSZmZn60Y9+pDlz5vjGPPzwwzp9+rSmTp2q48eP64YbbtDGjRuVnJzsG/Pyyy9r+vTpuuWWWxQfH68JEyZo6dKldkwJQBcFFFJoIqHV0feAgWjUs2dPLVmyREuWLGlzTFxcnObPn6/58+e3OSYtLU2rV68OQYUAwi2gkEITAUEKQChdrMfQd2IXv7sHAAAYiZACAACMREgBgBjAJ/kQiQgpESySm04k1w4ACA9CSgTq6n/whAMA0YwXQdEjoE/3IHZwNj2AriAkIBgIKQYiIACIVB0NJ/Q5dARv90QI7y8JBAAT8RYLQoGVlBhFMwHQEax4wE6spABADLnYigcrtTAVKykxxI7Vk+yK9XIkWFo0Juw3DQCIcIQUAEDATHjLmBdB0Y+QEgXaes/YhCYCILK01Tc4NwV2IKQAAIzECy1w4iwAADASISWK8D0FAIBoQkiBJAIOAMA8nJMSQQgRAIBYQkiJQoQZAKZ+Gof+hEAQUuCHBgLEJp77MBHnpAAAACOxkoIu4dUXEJk6+22tJjznTagB4UFIMRhf+QwgWhAs0BmEFIQEDQkA0FWck2IjvpsEAIC2sZISRqZ+JBBA9IrUF0KRWjeCi5UUAABgJFZSDBCJrxgisWYAQGRhJQUAokQkv3jgHD1cDCspAADbEExwKaykAAAAI7GSgg4J1asdPvEEBBcrE39Ff4l8rKQAAAAjEVIAAICReLsHABAReCsr9hBScEk0BQCAXQgpYcB/9AAABI6QYoNoDi3RPDfAVDzvEK04cRYAABiJkAIAAIxESAEAAEYipAAAACNx4iwARChOmEW0YyUFAAAYiZACAACMREgBAABGIqQAAAAjceJsCHFSW/B4f5aHFhbbXAkAIFxYSQEAAEYipAAAACMRUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBQAAGImQAiNkV6zne2UAAH4IKQAAwEiEFAAAYCRCCgAgqvF2cuQipAAAACMRUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBQAAGImQAgAAjERIAQAARiKkAAAAIxFSABjjD3/4g/7hH/5BvXv3Vvfu3TVs2DB98MEHvusty9KcOXPUr18/de/eXQUFBdq/f7/fMY4dO6aSkhI5nU6lpqZq8uTJOnXqVLinAiAIAg4pNBEAofCXv/xF119/vRITE/XWW29pz549+o//+A997Wtf841ZtGiRli5dquXLl6u+vl49evRQUVGRzp496xtTUlKi3bt3q6amRuvWrdOWLVs0depUO6YEoIu6BTLY20S+/e1v66233lKfPn20f//+izaRF198UTk5OZo9e7aKioq0Z88eJScnS/qyifzxj39UTU2N3G63Jk2apKlTp2r16tXBnR2AiPHkk08qKytLK1as8G3Lycnx/duyLC1ZskSPPvqoxo0bJ0latWqV0tPTtXbtWk2cOFF79+7Vxo0btWPHDo0ePVqStGzZMt1xxx166qmnlJmZGd5JAeiSgEIKTQRAqLz55psqKirS97//fb377rv6m7/5G91///2aMmWKJOngwYNyuVwqKCjw7dOrVy/l5eWprq5OEydOVF1dnVJTU329RZIKCgoUHx+v+vp63Xnnna1ut7m5Wc3Nzb7LTU1NkiS32y23291mvd7rLjUm1BwJVtf2j7f8/o5EgczBzvvqUkx4LHVVR+cQ6BwDCik0kcB0tYFINBHfMf7/Z2nXfWn3YykYQtVEguX3v/+9nnnmGZWXl+vf//3ftWPHDv3Lv/yLkpKSVFpaKpfLJUlKT0/32y89Pd13ncvlUt++ff2u79atm9LS0nxjLlRVVaXKyspW2zdv3qyUlJR2666pqenQ/EJh0ZjgHGfBaE9wDmSjjsxhw4YNYaik8+x8LAVLe3M4c+ZMQMcLKKTQRAITrAYi0US8P0u7mwxNJHQ8Ho9Gjx6tJ554QpI0cuRIffLJJ1q+fLlKS0tDdruzZs1SeXm573JTU5OysrJUWFgop9PZ5n5ut1s1NTW69dZblZiYGLL6vmrovE2SpE/mFfld7ixHvKUFoz2a/UG8mj1xXa7PDoHMwftzM40dj6Vg6+gcvIsMHRVQSKGJBKarDUSiiXhd2JzDze7HUjCEqokES79+/TR48GC/bYMGDdJ///d/S5IyMjIkSY2NjerXr59vTGNjo3Jzc31jjh496neM8+fP69ixY779L+RwOORwOFptT0xM7NB93dFxwdDcEue7za9e7vJxPXFBO5ZdOjIH05+74XwshUp7cwh0fgGFFJpIYIL5pI/1JnJhc7YLTSR0rr/+eu3bt89v229/+1sNGDBA0pfnv2VkZKi2ttbXT5qamlRfX69p06ZJkvLz83X8+HE1NDRo1KhRkqS3335bHo9HeXl54ZsMgKAI6CPIgTQRL28Tyc/Pl+TfRLxoIgAeeughbdu2TU888YQOHDig1atX67nnnlNZWZkkKS4uTjNmzNBjjz2mN998U7t27dK9996rzMxMjR8/XtKXL5puu+02TZkyRdu3b9f777+v6dOna+LEiZyUD0SggFZSHnroIV133XV64okndNddd2n79u167rnn9Nxzz0nybyJXX3217yPIbTWR5cuXy+12R00Tya5YL0k6tLDY5kqAyHPttddqzZo1mjVrlubPn6+cnBwtWbJEJSUlvjEPP/ywTp8+ralTp+r48eO64YYbtHHjRt/XG0jSyy+/rOnTp+uWW25RfHy8JkyYoKVLl9oxJQBdFFBIoYkACKXvfOc7+s53vtPm9XFxcZo/f77mz5/f5pi0tDS+cwmIEgGFFIkmAgAAwoPf3QMAAIwU8EoKAMBe3vPfgGjHSgoAADASKykhwKscAAC6jpUUAABgJEIKACAmZFesZ6U7whBSAACAkQgpMAqvdAAAXoQUAABgJEIKAAAwEh9BDgLengAAIPhYSQEAAEZiJQVG8q5OHVpYbHMlgP1YrUWsYiUFABBT+BRh5CCkAAAAIxFSAACAkQgpAADASIQUAABgJEIKAAAwEiEFAAAYie9JgdH4mCAAxC5WUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBQAAGImQAgCG4BffAf4IKQBgE0IJcGl8TwoA2IygAlwcKykAAMBIhBQAAGAkQgoAADASIQUAABiJkNIJnJEPAEDoEVICRDgBACA8CCkAAMBIhBQAMAxvKQNfIqQAAGISYdB8hBQAAGAkQgoAADASv7unC1gmBAAgdFhJAQAARmIlBQAMxWotYh0rKQAAwEiEFABATOOjyOYipAAAACMRUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBQAAGImQAgAAjERIAQAARuJr8TuIL/oBACC8WElBROGbIQEgdhBSAACAkQgpAADASIQUAABgJEIKAAAwEiEFAAAYiZACAID49KCJCCkAAMBIhBQAAGAkQgoAADASIQUAABiJkAIAAIxESAEAAEYipCAi8TFBAIh+hBQARlq4cKHi4uI0Y8YM37azZ8+qrKxMvXv31mWXXaYJEyaosbHRb7/Dhw+ruLhYKSkp6tu3r2bOnKnz58+HuXoAwdClkEITARAKO3bs0LPPPqvhw4f7bX/ooYf0i1/8Qq+99preffddHTlyRN/73vd817e0tKi4uFjnzp3T1q1b9eKLL2rlypWaM2dOuKeACMaXupmj0yGFJgIgFE6dOqWSkhL953/+p772ta/5tp84cULPP/+8Fi9erJtvvlmjRo3SihUrtHXrVm3btk2StHnzZu3Zs0cvvfSScnNzdfvtt2vBggWqrq7WuXPn7JoSgE7q1pmdvtpEHnvsMd92bxNZvXq1br75ZknSihUrNGjQIG3btk1jx471NZFf/vKXSk9PV25urhYsWKBHHnlE8+bNU1JSUnBmBiAilZWVqbi4WAUFBX79paGhQW63WwUFBb5tAwcOVP/+/VVXV6exY8eqrq5Ow4YNU3p6um9MUVGRpk2bpt27d2vkyJGtbq+5uVnNzc2+y01NTZIkt9stt9vdZp3e6y41pj2OBKvT+waDI97y+zsShXIOXblvO3M74bq9UOjoHAKdY6dCSiw1ES+7mglNpG3hfELTRMLjlVde0YcffqgdO3a0us7lcikpKUmpqal+29PT0+VyuXxjvtpbvNd7r7uYqqoqVVZWttq+efNmpaSktFtzTU1Nu2PasmhMp3cNqgWjPXaX0GWhmMOGDRuCfsxL6cpjyRTtzeHMmTMBHS/gkBJrTcTL7mZCE2kt3A1EoomE0meffaYHH3xQNTU1Sk5ODtvtzpo1S+Xl5b7LTU1NysrKUmFhoZxOZ5v7ud1u1dTU6NZbb1ViYmKnbnvovE2d2i9YHPGWFoz2aPYH8Wr2xNlaS2eFcg6fzCsK6vHaEozHkt06OgfvIkNHBRRSYrGJeNnVTGgibQtXA5FoIuHQ0NCgo0eP6pvf/KZvW0tLi7Zs2aKf/vSn2rRpk86dO6fjx4/7vRBqbGxURkaGJCkjI0Pbt2/3O673xH3vmAs5HA45HI5W2xMTEzt0X3d03MU0t5jxnG72xBlTS2eFYg7hfq535bFkivbmEOj8AgopsdhEvOx+AtNEWrPjyUwTCZ1bbrlFu3bt8ts2adIkDRw4UI888oiysrKUmJio2tpaTZgwQZK0b98+HT58WPn5+ZKk/Px8Pf744zp69Kj69u0r6cuVI6fTqcGDB4d3QgC6LKCQQhMBECo9e/bU0KFD/bb16NFDvXv39m2fPHmyysvLlZaWJqfTqQceeED5+fkaO3asJKmwsFCDBw/WPffco0WLFsnlcunRRx9VWVnZRV/oADBbQCGFJgLATj/5yU8UHx+vCRMmqLm5WUVFRXr66ad91yckJGjdunWaNm2a8vPz1aNHD5WWlmr+/Pk2Vg2gszr16Z5LoYkACJZf/epXfpeTk5NVXV2t6urqNvcZMGCALSdVAwi+LocUmggAAAgFfncPAAAwEiEFAAAYiZACAACMREgBAABGIqQAAAAjEVIAAICRCCkAAMBIhBQAAGAkQgoAADASIQUAABiJkAIAAIxESAEAAEYipAAAACMRUgAAuIjsivXKrlhvdxkxjZCCiEUDAYDoRkgBAABGIqQAAAAjEVIAAICRCCkAAMBIhBQAAGAkQgoAADASIQUAABiJkIKIx/elAAgleox9CCkAAMBIhBQAAGAkQgoAADASIQUAABiJkAIAQAdwAm34EVIQNWggABBdCCkAAMBIhBQAAGAkQgoAADBSN7sLMB3nOAAAYA9WUgAAgJFYSQGAMGOFFugYQsoFvM3j0MJimysBEG0IJ0BgCCltoJkAAGAvzkkBAABGIqQAAAAjEVIAAICRCCkAAMBIhBQAAGAkQgoAADASIQUAABiJkAIAAIxESAEAAEYipAAAACMRUgAAgJEIKQAAwEiEFAAAYCRCCgAAAciuWK/sivV2lxETutldgCl4wEUP7315aGGxzZUAX6K/AJ3DSgoAADASIQUAABiJkAIAAIxESAEAAEYipAAAACMRUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBQAAGImQAgAAjERIAQAARiKkAAAAIxFSAACAkQgpAADASIQUAABgJEIKACNUVVXp2muvVc+ePdW3b1+NHz9e+/bt8xtz9uxZlZWVqXfv3rrssss0YcIENTY2+o05fPiwiouLlZKSor59+2rmzJk6f/58OKcCIEgCCik0EQCh8u6776qsrEzbtm1TTU2N3G63CgsLdfr0ad+Yhx56SL/4xS/02muv6d1339WRI0f0ve99z3d9S0uLiouLde7cOW3dulUvvviiVq5cqTlz5tgxJQBdFFBIoYkACJWNGzfqvvvu05AhQzRixAitXLlShw8fVkNDgyTpxIkTev7557V48WLdfPPNGjVqlFasWKGtW7dq27ZtkqTNmzdrz549eumll5Sbm6vbb79dCxYsUHV1tc6dO2fn9AB0QrdABm/cuNHv8sqVK9W3b181NDToxhtv9DWR1atX6+abb5YkrVixQoMGDdK2bds0duxYXxP55S9/qfT0dOXm5mrBggV65JFHNG/ePCUlJbW63ebmZjU3N/suNzU1SZLcbrfcbneb9Xqvu9QYL0eC1f4PwAaOeMvv70hk1xw6cr8HeqxgHjPcOjoHU+Z44sQJSVJaWpokqaGhQW63WwUFBb4xAwcOVP/+/VVXV6exY8eqrq5Ow4YNU3p6um9MUVGRpk2bpt27d2vkyJGtbieU/WXovE2SJEdCu9O1Bf2la4L1XKG/tC2gkHKhcDWRqqoqVVZWttq+efNmpaSktFtnTU1Nu2MWjWl3iK0WjPbYXUKXhXsOGzZsCPoxO/JYMl17czhz5kyYKmmbx+PRjBkzdP3112vo0KGSJJfLpaSkJKWmpvqNTU9Pl8vl8o35am/xXu+97mJC2V9M7yte9JfOCXaPob+01umQEs4mMmvWLJWXl/suNzU1KSsrS4WFhXI6nW3W6Ha7VVNTo1tvvVWJiYmXnI/3FY9pHPGWFoz2aPYH8Wr2xNldTqfYPYdP5hV1+RiBPJZM1dE5eFcS7FRWVqZPPvlE7733XshvK5T9xdS+4mX3czMYTJhDV3sM/aVtnQ4p4WwiDodDDoej1fbExMQO3aEdGdfcYvYTtNkTZ3yN7bFrDsF80nf0MWey9uZg9/ymT5+udevWacuWLbriiit82zMyMnTu3DkdP37c74VQY2OjMjIyfGO2b9/udzzvifveMRcKZX+JlOcs/aVrrp69WZJ0aGFxl45Df2mtUx9B9jaRd955p80m8lUXNpELP+3TXhMJpeyK9cquWB/220X4cB9HBsuyNH36dK1Zs0Zvv/22cnJy/K4fNWqUEhMTVVtb69u2b98+HT58WPn5+ZKk/Px87dq1S0ePHvWNqampkdPp1ODBg8MzEQBBE1BIoYkACJWysjK99NJLWr16tXr27CmXyyWXy6UvvvhCktSrVy9NnjxZ5eXleuedd9TQ0KBJkyYpPz9fY8eOlSQVFhZq8ODBuueee/TrX/9amzZt0qOPPqqysrKLrpYAMFtAb/eUlZVp9erVeuONN3xNRPqyeXTv3t2viaSlpcnpdOqBBx5os4ksWrRILpeLJgJAzzzzjCTpW9/6lt/2FStW6L777pMk/eQnP1F8fLwmTJig5uZmFRUV6emnn/aNTUhI0Lp16zRt2jTl5+erR48eKi0t1fz588M1DQBBFFBIoYkACBXLav8jpMnJyaqurlZ1dXWbYwYMGBCST3YBCL+AQgpNBAAAhAu/uwcAABiJkAIAAIxESAEAAEYipAAAACMRUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBTGDXzQIIJQu7DH0m64jpCBmEVoAwGwBfS0+AKB9hF8gOFhJAQAARorZlRRe6QAAYDZWUgAAgJEIKQAAwEiEFAAAYCRCCgAAMBIhBQAAGImQAgAAjERIAWzCN94CCJVo6S+EFAAAYKSY/TI3AAi2aHjliq7jcRA8MRdSePCAxwAARAbe7gFCLJD3hqPlfWQA4RHtPYOQAkSA9hpRtDcqAMERab0i5t7uAaJJJDUbIBZ5n6OHFhaH9fba297Rur66X7jm8FWEFMS8YDWRzjzpO3rMQG+js+MBmClUz+W2+pEpPYOQAvw/O/9DbyuMAICpwtEzCSlAFxEoACA0CClAgHgLBUCgAn072Duuqy+CIv1FFCEFsFkwmwgBCogNkR4+OoqQAlwg3CeoAYgdoTjBvjPjO3M8O14EEVKAAAydt0lSnCR7QweBB0AsIKQAQUaAAIDgIKQAMSC7Yr0cCZYWjbG7EgBt+epKbVti7UUQX4sPAACMREgBACDMIu136NiFt3uANvBxXgBoXyjfTmYlBQAAGImQAgAAjMTbPUAU4r1uANGAkAIAgE0i8esBwvkiiJACtCMSmwgARAPOSQEAAEYipAAAACMRUgAAgJE4JwUAuohPUwGhwUoKAAAwEiEFAAAYiZACAACMFDPnpPCeMQAAkYWVFAAAYCRCCgAAMBIhBQAAGImQAgAAjERIAQAARiKkAAAAIxFSAACAkQgpAADASIQUAABgJEIKAAAwEiEFAAAYiZACAACMREgBAABGIqQAAAAjEVIAAICRCCkAAMBIMRFShs7bZHcJAKIU/QUIHVtDSnV1tbKzs5WcnKy8vDxt377dznIARBH6CxD5bAspr776qsrLyzV37lx9+OGHGjFihIqKinT06FG7SgIQJegvQHSwLaQsXrxYU6ZM0aRJkzR48GAtX75cKSkpeuGFF+wqCUCUoL8A0aGbHTd67tw5NTQ0aNasWb5t8fHxKigoUF1dXavxzc3Nam5u9l0+ceKEJOnYsWNyu91t3o7b7daZM2fUzR2vFk9cEGcQPt08ls6c8TAHm0XTHP785z8rMTGxzXEnT56UJFmWFa7Sgor+0nHR9LhmDvYKVX+xJaR8/vnnamlpUXp6ut/29PR0/eY3v2k1vqqqSpWVla225+TkhKxGk/zQ7gKCgDmYIZA5nDx5Ur169QpZLaFCfwlMrD2uTRVrc+hof7ElpARq1qxZKi8v9132eDw6duyYevfurbi4tlNnU1OTsrKy9Nlnn8npdIaj1KBjDmaIpTlYlqWTJ08qMzMzjNXZh/7CHOwWS3MItL/YElIuv/xyJSQkqLGx0W97Y2OjMjIyWo13OBxyOBx+21JTUzt8e06nM2LveC/mYIZYmUMkrqB40V8CxxzMECtzCKS/2HLibFJSkkaNGqXa2lrfNo/Ho9raWuXn59tREoAoQX8Boodtb/eUl5ertLRUo0eP1pgxY7RkyRKdPn1akyZNsqskAFGC/gJEB9tCyg9+8AP96U9/0pw5c+RyuZSbm6uNGze2OtmtKxwOh+bOndtqKTeSMAczMIfIQn/pGOZgBubQtjgrUj9nCAAAolpM/O4eAAAQeQgpAADASIQUAABgJEIKAAAwEiEFAAAYKepCyuOPP67rrrtOKSkpHf7WSMuyNGfOHPXr10/du3dXQUGB9u/fH9pCL+HYsWMqKSmR0+lUamqqJk+erFOnTl1yn29961uKi4vz+/PP//zPYapYqq6uVnZ2tpKTk5WXl6ft27dfcvxrr72mgQMHKjk5WcOGDdOGDRvCVGnbApnDypUrW/28k5OTw1hta1u2bNF3v/tdZWZmKi4uTmvXrm13n1/96lf65je/KYfDoa9//etauXJlyOuMZPQX+ktn0V8611+iLqScO3dO3//+9zVt2rQO77No0SItXbpUy5cvV319vXr06KGioiKdPXs2hJW2raSkRLt371ZNTY3WrVunLVu2aOrUqe3uN2XKFP3xj3/0/Vm0aFEYqpVeffVVlZeXa+7cufrwww81YsQIFRUV6ejRoxcdv3XrVt19992aPHmydu7cqfHjx2v8+PH65JNPwlLvxQQ6B+nLr3/+6s/7008/DWPFrZ0+fVojRoxQdXV1h8YfPHhQxcXF+va3v62PPvpIM2bM0D/90z9p06ZNIa40ctFf6C+dQX/pQn+xotSKFSusXr16tTvO4/FYGRkZ1o9//GPftuPHj1sOh8P62c9+FsIKL27Pnj2WJGvHjh2+bW+99ZYVFxdn/eEPf2hzv5tuusl68MEHw1Bha2PGjLHKysp8l1taWqzMzEyrqqrqouPvuusuq7i42G9bXl6e9aMf/SikdV5KoHPo6OPLLpKsNWvWXHLMww8/bA0ZMsRv2w9+8AOrqKgohJVFB/pL+NBfzBPO/hJ1KymBOnjwoFwulwoKCnzbevXqpby8PNXV1YW9nrq6OqWmpmr06NG+bQUFBYqPj1d9ff0l93355Zd1+eWXa+jQoZo1a5bOnDkT6nJ17tw5NTQ0+P384uPjVVBQ0ObPr66uzm+8JBUVFdny85Y6NwdJOnXqlAYMGKCsrCyNGzdOu3fvDke5QWPa/RCN6C9dQ3+hv9j2tfimcLlcktTq67LT09N914W7nr59+/pt69atm9LS0i5Zzw9/+EMNGDBAmZmZ+vjjj/XII49o3759ev3110Na7+eff66WlpaL/vx+85vfXHQfl8tlzM9b6twcrrnmGr3wwgsaPny4Tpw4oaeeekrXXXeddu/erSuuuCIcZXdZW/dDU1OTvvjiC3Xv3t2myqIH/aVr6C/0l4hYSamoqGh1EtGFf9q6s00R6jlMnTpVRUVFGjZsmEpKSrRq1SqtWbNGv/vd74I4C3jl5+fr3nvvVW5urm666Sa9/vrr6tOnj5599lm7S0OA6C/to7+EF/3lryJiJeVf//Vfdd99911yzJVXXtmpY2dkZEiSGhsb1a9fP9/2xsZG5ebmduqYF9PROWRkZLQ6mer8+fM6duyYr9aOyMvLkyQdOHBAV111VcD1dtTll1+uhIQENTY2+m1vbGxss96MjIyAxodaZ+ZwocTERI0cOVIHDhwIRYkh0db94HQ6Y2oVhf5Cfwkl+stfdaa/RERI6dOnj/r06ROSY+fk5CgjI0O1tbW+ptHU1KT6+vqAzuBvT0fnkJ+fr+PHj6uhoUGjRo2SJL399tvyeDy+xtARH330kST5NcZQSEpK0qhRo1RbW6vx48dLkjwej2prazV9+vSL7pOfn6/a2lrNmDHDt62mpkb5+fkhrbUtnZnDhVpaWrRr1y7dcccdIaw0uPLz81t9NNPO+8Eu9Bf6SyjRX/6qU/dDoGf1mu7TTz+1du7caVVWVlqXXXaZtXPnTmvnzp3WyZMnfWOuueYa6/XXX/ddXrhwoZWammq98cYb1scff2yNGzfOysnJsb744gs7pmDddttt1siRI636+nrrvffes66++mrr7rvv9l3/v//7v9Y111xj1dfXW5ZlWQcOHLDmz59vffDBB9bBgwetN954w7ryyiutG2+8MSz1vvLKK5bD4bBWrlxp7dmzx5o6daqVmppquVwuy7Is65577rEqKip8499//32rW7du1lNPPWXt3bvXmjt3rpWYmGjt2rUrLPVeTKBzqKystDZt2mT97ne/sxoaGqyJEydaycnJ1u7du+2agnXy5Enf412StXjxYmvnzp3Wp59+almWZVVUVFj33HOPb/zvf/97KyUlxZo5c6a1d+9eq7q62kpISLA2btxo1xSMR3+hv3QG/aXz/SXqQkppaaklqdWfd955xzdGkrVixQrfZY/HY82ePdtKT0+3HA6Hdcstt1j79u0Lf/H/789//rN19913W5dddpnldDqtSZMm+TXBgwcP+s3p8OHD1o033milpaVZDofD+vrXv27NnDnTOnHiRNhqXrZsmdW/f38rKSnJGjNmjLVt2zbfdTfddJNVWlrqN/7nP/+59Y1vfMNKSkqyhgwZYq1fvz5stbYlkDnMmDHDNzY9Pd264447rA8//NCGqv/qnXfeuehj31t3aWmpddNNN7XaJzc310pKSrKuvPJKv+cFWqO/0F86i/7Suf4SZ1mWFfA6DgAAQIhFxKd7AABA7CGkAAAAIxFSAACAkQgpAADASIQUAABgJEIKAAAwEiEFAAAYiZACAACMREgBAABGIqQAAAAjEVIAAICR/g98g7jFJ5b/+QAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "res[['b','d']].hist(bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
